source_ds9/mm_comms9.c

```diff
@@ -0,0 +1,500 @@
+    SendSimpleExt(module_ID, mode, MSG_START);
+}
+
+// Pause module playback
+void mmPause() {
+    SendSimple(0, MSG_PAUSE);
+}
+
+// Resume module playback
+void mmResume() {
+    SendSimple(0, MSG_RESUME);
+}
+
+// Stop module playback
+void mmStop() {
+    SendSimple(0, MSG_STOP);
+}
+
+// Set playback position
+void mmPosition(mm_word position) {
+    SendSimple(position, MSG_POSITION);
+}
+
+// Start jingle
+void mmJingle(mm_word module_ID) {
+    SendSimple(module_ID, MSG_STARTSUB);
+}
+
+// Set module volume
+void mmSetModuleVolume(mm_word vol) {
+    SendSimple(vol, MSG_MASTERVOL);
+}
+
+// Set jingle volume
+void mmSetJingleVolume(mm_word vol) {
+    SendSimple(vol, MSG_MASTERVOLSUB);
+}
+
+// Set master tempo
+void mmSetModuleTempo(mm_word tempo) {
+    SendSimple(tempo, MSG_MASTERTEMPO);
+}
+
+// Set master pitch
+void mmSetModulePitch(mm_word pitch) {
+    SendSimple(pitch, MSG_MASTERPITCH);
+}
+
+// Set master effect volume
+void mmSetEffectsVolume(mm_word vol) {
+    SendSimple(vol, MSG_MASTEREFFECTVOL);
+}
+
+// Select audio mode
+void mmSelectMode(mm_mode_enum mode) {
+    SendSimple(mode, MSG_SELECTMODE);
+}
+
+// Open audio stream
+void mmStreamBegin(mm_word wave, mm_hword clks, mm_hword len, mm_byte format) {
+    mm_word buffer[MAX_PARAM_WORDS];
+    
+    buffer[0] = (wave << 16) | (MSG_OPENSTREAM << 8) | (10);
+    buffer[1] = (wave >> 16) | (clks << 16);
+    buffer[2] = len | (format << 16);
+    
+    SendString(buffer, 3);
+}
+
+// Close audio stream
+void mmStreamEnd() {
+    SendSimple(0, MSG_CLOSESTREAM);
+}
+
+// Returns same handle, or a newer valid handle
+static mm_sfxhand mmValidateEffectHandle(mm_sfxhand handle) {
+    mm_byte instance_num = (handle & 0xFF) - 1;
+
+    if((instance_num < EFFECT_CHANNELS) && (sfx_instances[instance_num] == ((handle >> 8) & 0xFF)))
+        return handle;
+    
+    return mmCreateEffectHandle();
+}
+
+// Return effect handle
+// NO_HANDLES_AVAILABLE = no channels available
+static mm_sfxhand mmCreateEffectHandle() {
+    int i = 0;
+    
+    // Search for free channel
+    for(; i < EFFECT_CHANNELS; i++)
+        if(!((sfx_bitmask>>i) & 1))
+            break;
+    
+    if(i >= EFFECT_CHANNELS)
+        return NO_HANDLES_AVAILABLE;
+
+    // Disable IRQ, cannot be interrupted by sfx update
+    mm_hword old_ime = REG_IME;
+    REG_IME = 0;
+    
+    // Set bit
+    sfx_bitmask |= 1 << i;
+    
+    // Re-enable IRQ
+    REG_IME = old_ime;
+
+    sfx_instances[i] += 1;
+    
+    return (sfx_instances[i] << 8) | (i + 1);
+}
+
+// Play sound effect, default parameters
+mm_sfxhand mmEffect(mm_word sample_ID) {
+    mm_word buffer[MAX_PARAM_WORDS];
+
+    mm_sfxhand handle = mmCreateEffectHandle();
+    
+    if(handle != NO_HANDLES_AVAILABLE) {
+        buffer[0] = (sample_ID << 16) | (MSG_EFFECT << 8) | (5);
+        buffer[1] = handle;
+        
+        SendString(buffer, 2);
+    }
+
+    return handle;
+}
+
+// Set effect volume
+// volume should be a byte... :/
```

Comment:
Does it actually matter that it is a word? As long as the documentation is clear about the valid range, it's ok, I think.

Reply:
Yeah, it's just me being pedantic, tbh.